enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,432 @@
+
+While not required initially, other common SLO patterns can be
+considered in future:
+
+1. Implement an operator configuration resource, including
+   `OperatorSpec` and `OperatorStatus` (as per in openshift/api#125)
+2. Implement cluster-level configuration under `config.openshift.io`
+   (as per openshift/api#124)
+3. Expose a `/metrics` endpoint for Prometheus to be configured to
+   scrape (via a `ServiceMontitor`) and define any relevant Prometheus
+   alert rules based on those metrics.
+
+### "Not In Use" SLO Behaviors
+
+Unlike most other SLOs, CBO is not applicable to all cluster
+configurations. On clusters running on an infrastructure platform
+other than `baremetal` it should adhere to the emerging expected
+behaviors for "not in use" SLOs, including:
+
+1. Setting its `ClusterOperator` condition to `Disabled=true`,
+  `Available=true`, `Progressing=False` with appropriate messages.
+2. User interfaces should convey this disabled state differently than
+   a failure mode (e.g. by graying it out).
+
+Currently, insights-operator is the only other example of an SLO
+following this pattern of using a `Disabled` status. Other somewhat
+similar cases following different patterns include:
+
+- Image registry and samples have a `Removed` management state where
+  `Degraded=False`, `Progressing=False`, `Available=True` with
+  `Reason=currentlyUnmanaged`.
+- The cluster credentials operator has a `disabled` config map setting
+  that can be used to disable the operator and it then sets
+  ClusterOperator status conditions to `Degraded=False`,
+  `Progressing=False`, `Available=True` with
+  `Reason=OperatorDisabledByAdmin` for all three conditions.
+
+### cluster-baremetal-operator Details
+
+The CBO will:
+
+- Be a new `openshift/cluster-baremetal-operator` project.
+- Publish an image called `cluster-baremetal-operator`.
+- Use CVO run-level 30, so its manifests will be applied in parallel
+  with the MAO.
+- Add a new `baremetal` `ClusterOperator` with an additional
+  `Disabled` status for non-baremetal platforms.
+- Use the existing `openshift-machine-api` namespace where the
+  BareMetalHost resources are also located.
+- Install the `metal3.io` `Provisioning` (cluster-scoped) and
+  `BareMetalHost` (namespaced) CRDs.
+- Run under a new `openshift-machine-api/cluster-baremetal-operator`
+  `ServiceAccount`.
+- Be launched by a `openshift-machine-api/cluster-baremetal-operator`
+  `Deployment`, copying much of the MAO pod spec in terms of
+  `system-node-critical` priority class, running on masters, security
+  context, resource requests, etc.
+- Implement a controller reconciling the singleton
+  `provisioning-configuration` cluster-scoped `Provisioning` resource
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource a platform
+  type other than `BareMetal`.
+- Based on the values in the `Provisioning` resource, create a
```

Comment:
Yes, in terms of the main responsibility we're taking over from MAO - i.e. managing the `metal3` deployment - that is almost entirely controlled by the `Provisioning` resource. I'm pretty sure most or all of the data in that resource ends up as env vars for the containers in the deployment.

The only other resource influencing this is the `status.platform` field in `Infrastructure` resource.

So in the [initial implementation](https://github.com/markmc/cluster-baremetal-operator) all this code is in a controller for the provisioning resource

It might be a bit strong to say the controller is "reconciling" this resource though - the assumption in the MAO code for this seems to be that this is install-time configuration data which can't be changed

Reply:
Added a note on this